### PR TITLE
Priority-based leadership rebalancing for CP subsystem [HZ-1062]

### PR DESCRIPTION
**Pull request by arodionov**
The following commit introduces the priority-based leadership rebalancing in addition to the evenly rebalancing strategy.

Hazelcast CP Subsystem is based on the Raft protocol, which is a leader-based protocol.
A leader is responsible for client interaction and state replication.
Every CP group has its own leader and any CP member could be elected as a CP group leader.
To have an equal CP members load, the periodically running leadership rebalancing task is trying to distribute CP group leaders among all CP members evenly.
However, there are cases/scenarios when some CP members shouldn't act as a leader. To ensure the availability of the CP subsystem, we cannot prevent some members from being elected as a leader. But we can ask to transfer leadership from the existing leader to any other CP member.
To implement this behaviour, we can introduce priority-based leadership in addition to the existing rebalancing strategy.
Each CP member will be assigned a priority - an integer number. Initially, any CP member with any priority can take the leader's role. But the periodically running leadership rebalancing task should transfer the leadership to members with the highest priority.
So, eventually, all CP groups' leaders will be transferred to CP members with the highest priority.

To use the priority-based leadership strategy, just assign a priority to CP members:
`config.getCPSubsystemConfig().setCPMemberPriority(-1);`

After applying this configuration, every 60 sec (specified in the `LEADERSHIP_BALANCE_TASK_PERIOD` property) will run a `RaftGroupLeadershipBalanceTask`.

If the CP group size is smaller than the number of CP members, the groups may have different sets of CP members, and the highest CP member in each group may not be the same. In that case, we form a set of highest priorities based on the highest priorities taken from each group. For example:
_Group1: {M0-0; M1-1; M2-2} - max priority member is M2 with priority 2_
_Group2: {M1-1; M2-2; M3-3} - max priority member is M3 with priority 3_
The highest priorities set is: _{2, 3}_ 
Based on this highest priority set, we assume all CP members with those priorities are eligible for leadership.

Limitations:
In this PR, we do not address the problem where a CP group can only be formed from CP members with the lowest priority, or later on promoting CP members with the highest priority. To fix this, we should change the logic of how the CP groups are formed. Possibly we should introduce so-called _Light CP Members_. 

Fixes https://hazelcast.atlassian.net/browse/HZ-651

Breaking changes (list specific methods/types/messages):
- the return type for `GetLeadedGroupsOp` has been changed, but we can ignore it since the class cast exception will be caught [here](https://github.com/hazelcast/hazelcast/blob/14a6eec2cd682ad365e03907ad22dfc5da6cb866/hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java#L462) and rebalancing will be skipped during the rolling upgrade. 

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java [hunk 1]
```diff
@@ -464,8 +469,93 @@ private void rebalanceLeaderships() {
...
+            Map<CPMember, Collection<CPGroupId>> leaderships
+                    = rebalanceLeadershipsByPriority(leadershipPriorityMap, members, groupIds);
+
+            // If there are several CP members with equal priorities, we should rebalance leadership evenly between them
+            if (leaderships.size() > 1) {
+                rebalanceLeadershipsEvenly(leaderships, groupIds, members.values());
+            }
+        }
+
+        private Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>> getLeadershipPriorityMap(Collection<CPMember> members) {
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:484]: can we use some domain-specific classes instead of relying on generics here? `Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>>` is super un-intuitive what it actually contains

Comment (arodionov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:484]: Thanks Tomasz, I agree with your comment.
I think there is no sense to create a specific DTO for this case.
The best option will be to use [BiTuple](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/util/BiTuple.java), but this class is not serializable. If we make it serializable, we should make the other tuples serializable as well.

**Comment by arodionov** [comment]
Have found an uncovered scenario in the proposed priority-based leadership rebalancing implementation.
In general, we can have as many CP members as we want (for example 5).
The algorithm elects a separate leader per each CPGroup, that contains a random set of CP members (for example 3 members): https://github.com/arodionov/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/cp/internal/RaftInvocationManager.java#L137
In that case, the CP member(s) with the highest priority could not be part of CPGroup and the leadership transfer fails.

**Comment by arodionov** [comment]
@ramizdundar @ufukyilmaz To address the previous comment https://github.com/hazelcast/hazelcast/pull/21729#issuecomment-1184256339 the following implementation has been proposed:

If the CP group size is smaller than the number of CP members, the groups may have different sets of CP members, and the highest CP member in each group may not be the same. In that case, we form a set of highest priorities based on the highest priorities taken from each group. For example:
_Group1: {M0-0; M1-1; M2-2} - max priority member is M2 with priority 2_
_Group2: {M1-1; M2-2; M3-3} - max priority member is M3 with priority 3_
The highest priorities set is: _{2, 3}_ 
Based on this highest priority set we assume that all CP members with those priorities are eligible for leadership.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java [hunk 2]
```diff
@@ -461,25 +470,136 @@ public void run() {
...
+                    = groupIds.stream()
+                    .map(this::getCpGroup)
+                    .collect(Collectors.toList());
+
+            // Transferring leaderships from the lowest priority CP members
+            Set<CPMember> priorityCPMembers
+                    = rebalanceLeadershipsByPriority(leadershipPriorityMap, cpGroupSummaries);
+
+            // If there are several CP members with the same priority level, we should rebalance leadership evenly among them
+            if (priorityCPMembers.size() > 1) {
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:488]: I think this condition removed. For example:

- I have 5 members, 2 have 99 and 3 have 0 priority.
- I promote 2 new members with 99 priority.
- I would expect my groups to be rebalanced evenly.

Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:488]: Maybe we can have a test case for this scenario.

**Review by ramizdundar — COMMENTED** [review]
Whoa, what a PR. Those streams and generics punch hard. Good work!

File: hazelcast/src/main/java/com/hazelcast/config/cp/CPSubsystemConfig.java [hunk 3]
```diff
@@ -279,6 +279,12 @@ public class CPSubsystemConfig {
      */
     private int dataLoadTimeoutSeconds = DEFAULT_DATA_LOAD_TIMEOUT_SECONDS;
 
+    /**
+     * The CP member priority. The Raft leader's role eventually will be transferred to members with the
+     * highest priority.
+     */
+    private int cpMemberPriority;
```
[the remark this answers is not in the document]
Comment (arodionov) [reply hazelcast/src/main/java/com/hazelcast/config/cp/CPSubsystemConfig.java:286]: Yes, have a separate task for this https://hazelcast.atlassian.net/browse/HZ-1064
And I was planning to implement all other configurations as a separate PR after the "core" functionality will be merged.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java (hunk 2 above)
Comment (arodionov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:488]: Thanks, @ramizdundar for your comments!
Unfortunately, the current implementation could only rebalance leadership within the existing CP group members.  With the existing API, there is no mechanism to replace one CP member with another one. The only way to change CP group members is to remove the CP member https://docs.hazelcast.com/hazelcast/5.1/cp-subsystem/management#removing-a-cp-member and promote a new one.

My suggestion is don’t change the CP group members' pre-selection logic before the release, and leave the decision to the next iteration/release.

To summarise the limitations of the current implementation:
- a CP group can only be formed from CP members with the lowest priority
- new CP members (promoted members) with the highest priority will not be counted for existing groups when rebalancing

File: hazelcast/src/main/java/com/hazelcast/config/cp/CPSubsystemConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (arodionov) [reply hazelcast/src/main/java/com/hazelcast/config/cp/CPSubsystemConfig.java:286]: Here is a PR for configuration changes https://github.com/hazelcast/hazelcast/pull/21941

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java (hunk 2 above)
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:488]: Thanks for the answer.

**Review by ramizdundar — APPROVED** [review]

**Comment by arodionov** [comment]
run-lab-run

**Comment by arodionov** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java [hunk 4]
```diff
@@ -461,25 +470,136 @@ public void run() {
...
+
+        private Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>> getLeadershipPriorityMap(Collection<CPMember> members) {
+            Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>> leadershipsWithPriorities = new HashMap<>();
+            OperationService operationService = nodeEngine.getOperationService();
+            StringBuilder s = new StringBuilder("Current leadership claims:");
+            for (CPMember member : members) {
+                InvocationFuture<Map.Entry<Integer, Collection<CPGroupId>>> invocationFuture =
+                        operationService.invokeOnTarget(RaftService.SERVICE_NAME, new GetLeadedGroupsOp(),
+                                member.getAddress());
+                if (!invocationFuture.isCompletedExceptionally()) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java:501]: how can we know if the future is completed exceptionally before it is not done? Shouldn't we check `isDone()` first?

**Comment by ahmetmircik** [comment]
How can one see this leadership transfer campaign is finished? Is there any way to listen/monitor it?

**Comment by arodionov** [comment]
> How can one see this leadership transfer campaign is finished? Is there any way to listen/monitor it?

The leadership transfer only triggers the new Raft leader's election, so it can take some time. In theory, there is an API where we can ask if we have a leader. 
If leadership transfer fails (for example, the target member is unavailable or is not up-to-date to take the leader's role) we will retry to do rebalancing next time after 60 sec.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java (hunk 4 above)
Comment (arodionov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java]: Thanks, @ahmetmircik for the comment!
Have refactored this part of the code with a `try-catch` block.

**Comment by ahmetmircik** [comment]
Regarding this comment: https://github.com/hazelcast/hazelcast/pull/21729#issuecomment-1217779738

Do you know is there a way for users to monitor current members of a cp-group, maybe over MC? This can be helpful for a user to see if priorities really work. Though it is not directly in scope of this PR.

**Review by ahmetmircik — APPROVED** [review]

**Comment by arodionov** [comment]
Regarding this [comment](https://github.com/hazelcast/hazelcast/pull/21729#issuecomment-1218022543) 

There are logs that show how rebalancing is going on:
```
15:57:37,125  INFO |testRebalancingWithTwoNonPriorityMembers| - [RaftGroupMembershipManager] testRebalancingWithTwoNonPriorityMembers - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Current leadership claims:
	CPMember{uuid=a46b305a-5775-4eee-a7ba-71ad84541e79, address=[127.0.0.1]:5704} priority -1 has 6,
	CPMember{uuid=7a9a2f58-14bb-4fc7-95c9-c9b32d9d6bb8, address=[127.0.0.1]:5705} priority -1 has 11,
	CPMember{uuid=dd5b9aca-89d9-4118-9306-73d252ef469f, address=[127.0.0.1]:5701} priority 0 has 11,
	CPMember{uuid=83f324d5-c046-411c-8b79-12c14b0c7766, address=[127.0.0.1]:5702} priority 0 has 7,
	CPMember{uuid=a4a78348-24c0-4c2a-b517-4ea9b7bb0a4e, address=[127.0.0.1]:5703} priority 0 has 10, leaderships.
...
15:57:37,179  INFO |testRebalancingWithTwoNonPriorityMembers| - [RaftGroupMembershipManager] testRebalancingWithTwoNonPriorityMembers - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Current leadership claims:
	CPMember{uuid=a46b305a-5775-4eee-a7ba-71ad84541e79, address=[127.0.0.1]:5704} priority -1 has 0,
	CPMember{uuid=7a9a2f58-14bb-4fc7-95c9-c9b32d9d6bb8, address=[127.0.0.1]:5705} priority -1 has 0,
	CPMember{uuid=dd5b9aca-89d9-4118-9306-73d252ef469f, address=[127.0.0.1]:5701} priority 0 has 15,
	CPMember{uuid=83f324d5-c046-411c-8b79-12c14b0c7766, address=[127.0.0.1]:5702} priority 0 has 15,
	CPMember{uuid=a4a78348-24c0-4c2a-b517-4ea9b7bb0a4e, address=[127.0.0.1]:5703} priority 0 has 15, leaderships.
```

Also in MC, it is possible to see CP members of each group:
image

Maybe it will make sense to add a CP member priority column here.

**Comment by ahmetmircik** [comment]
@arodionov Good to know that, thanks.
And definitely that will be useful to see priorities on MC.
